From 64b31af9ab5d8212b77c262406843501e8373639 Mon Sep 17 00:00:00 2001
From: smomin7 <smomin@vicon-security.com>
Date: Thu, 24 Apr 2025 19:04:55 +0530
Subject: [PATCH 17/18] gstaggregator: Update include following commits commit:
 447da9a * aggregator: handle erroneous case of the first buffer before
 segment Segment event always should arrive before the first buffer, however
 if that happens due to some complex pipelines manipulations, we don't want
 the process to crash or go into errouneous state. In partigular an abort gets
 raised by the videoaggregator in this case. To avoid that we propose to
 safely assume we have received a segment in which the first buffer is valid.

---
 .../gstreamer/libs/gst/base/gstaggregator.c   | 38 +++++++++++--------
 1 file changed, 22 insertions(+), 16 deletions(-)

diff --git a/subprojects/gstreamer/libs/gst/base/gstaggregator.c b/subprojects/gstreamer/libs/gst/base/gstaggregator.c
index b2d6fcbe6f..193982126c 100644
--- a/subprojects/gstreamer/libs/gst/base/gstaggregator.c
+++ b/subprojects/gstreamer/libs/gst/base/gstaggregator.c
@@ -1824,6 +1824,9 @@ gst_aggregator_default_sink_event (GstAggregator * self,
         SRC_LOCK (self);
         priv->send_eos = TRUE;
         priv->got_eos_event = FALSE;
+        if (self->priv->start_time_selection ==
+            GST_AGGREGATOR_START_TIME_SELECTION_FIRST)
+          priv->first_buffer = TRUE;
         SRC_BROADCAST (self);
         SRC_UNLOCK (self);
 
@@ -3365,23 +3368,26 @@ gst_aggregator_pad_chain_internal (GstAggregator * self,
         break;
       case GST_AGGREGATOR_START_TIME_SELECTION_FIRST:
         GST_OBJECT_LOCK (aggpad);
-        if (aggpad->priv->head_segment.format == GST_FORMAT_TIME) {
-          start_time = buf_pts;
-          if (start_time != -1) {
-            if (aggpad->priv->head_segment.rate < 0.0 && buf_duration != -1) {
-              start_time += buf_duration;
-            }
-            start_time = MAX (start_time, aggpad->priv->head_segment.start);
-            start_time =
-                gst_segment_to_running_time (&aggpad->priv->head_segment,
-                GST_FORMAT_TIME, start_time);
-          }
-        } else {
-          start_time = 0;
+        if (aggpad->priv->head_segment.format != GST_FORMAT_TIME) {
+          gst_segment_init (&aggpad->priv->head_segment, GST_FORMAT_TIME);
+          gst_segment_do_seek (&aggpad->priv->head_segment, 1.0, GST_FORMAT_TIME,
+            GST_SEEK_FLAG_ACCURATE, GST_SEEK_TYPE_SET, buf_pts,
+            GST_SEEK_TYPE_SET, GST_CLOCK_TIME_NONE, NULL);
           GST_WARNING_OBJECT (aggpad,
-              "Ignoring request of selecting the first start time "
-              "as the segment is a %s segment instead of a time segment",
-              gst_format_get_name (aggpad->priv->head_segment.format));
+            "No valid segment received before the first buffer. Assuming a segment "
+            "from the buffer pts: %" GST_SEGMENT_FORMAT, &aggpad->priv->head_segment);
+        }
+
+        start_time = buf_pts;
+        if (start_time != -1) {
+          //GstClockTime buf_duration = GST_BUFFER_DURATION (buffer);
+          if (aggpad->priv->head_segment.rate < 0.0 && buf_duration != -1) {
+            start_time += buf_duration;
+          }
+          start_time = MAX(start_time, aggpad->priv->head_segment.start);
+          start_time =
+            gst_segment_to_running_time(&aggpad->priv->head_segment,
+              GST_FORMAT_TIME, start_time);
         }
         GST_OBJECT_UNLOCK (aggpad);
         break;
-- 
2.40.1.windows.1

